hack/test-go.sh

```diff
@@ -77,22 +77,45 @@ fi
 
 KUBE_TIMEOUT=${KUBE_TIMEOUT:--timeout 60s}
 
-if [ "${1-}" != "" ]; then
-  if [[ "${1}" == *"/..." ]]; then
+# Declare and set arrays to null string. We need to always keep the first
+# element to be the null string, otherwise the variable is considered unset.
+# When using, slice it to ignore the first element: ${varname[@]:1}.
+declare -a test_flags= test_packages=
+
+if [ $# -gt 1 ]; then
+  for arg in "$@"; do
+    # "." might be used in "-bench ." test flag
+    if [ "$arg" = '.' ]; then
+      test_flags[${#test_flags[@]}]="${arg}"
+
     # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1:0:(-4)}")
+    elif [[ "$arg" == *"/..." ]]; then
+      test_packages+=( $(find_test_dirs "${arg:0:(-4)}") )
+
+    # if arg starts with "github.com/openshift/origin/", append to test_packages
+    elif [ -z "${arg/#"${OS_GO_PACKAGE}/"*/}" ]; then
+      test_packages[${#test_packages[@]}]="${arg}"
+
+    # if arg is a directory, e.g., "pkg/build/api", prefix it with
+    # "github.com/openshift/origin/" and append to test_packages
+    elif [ -d "$arg" ]; then
+      test_packages[${#test_packages[@]}]="$OS_GO_PACKAGE/$arg"
+
+    # default: use argument as test flag
+    else
+      test_flags[${#test_flags[@]}]="${arg}"
+    fi
+  done
+else
+  if [ -n "$TEST_KUBE" ]; then
+    test_packages+=( `find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs` )
   else
-    test_packages="$OS_GO_PACKAGE/$1"
+    test_packages+=( `find_test_dirs "."; special_upstream_test_dirs` )
```

Comment:
If we're not using `$1` anymore here, why is the above `if` statement hinging on `[ $# -gt 1 ]` ?